python/tests/test_umap.py

```diff
@@ -672,4 +675,97 @@ def test_umap_sparse_vector(
         loc_umap = trustworthiness(input_raw_data.toarray(), embedding, n_neighbors=15)
 
         trust_diff = loc_umap - dist_umap
-        assert trust_diff <= 0.15
+
+        assert trust_diff <= 0.07
+
+
+@pytest.mark.parametrize("knn_graph_format", ["sparse", "dense", "tuple"])
+def test_umap_precomputed_knn(
+    knn_graph_format: str, tmp_path: str, gpu_number: int
+) -> None:
+
+    n_rows = 10000
+    n_cols = 50
+    random_state = 42
+
+    X, _ = make_blobs(
+        n_rows,
+        n_cols,
+        centers=5,
+        cluster_std=0.1,
+        dtype=cp.float32,
+        random_state=random_state,
+    )
+
+    k = 15
+    knn_metric = "sqeuclidean"
+
+    # Test a few different KNN implementations
+    if knn_graph_format == "tuple":
+        from cuvs.neighbors import cagra
+
+        X_row_major = cp.ascontiguousarray(X)
+        build_params = cagra.IndexParams(metric=knn_metric)
+        index = cagra.build(build_params, X_row_major)
+        distances, neighbors = cagra.search(cagra.SearchParams(), index, X_row_major, k)
+        distances = cp.asarray(distances)
+        neighbors = cp.asarray(neighbors)
+        precomputed_knn = (neighbors, distances)
+    elif knn_graph_format == "sparse":
+        from cuml.neighbors import NearestNeighbors
+
+        knn_model = NearestNeighbors(n_neighbors=k, metric=knn_metric)
+        knn_model.fit(X)
+        precomputed_knn = knn_model.kneighbors_graph(X)
+        assert isinstance(precomputed_knn, cupyx.scipy.sparse.spmatrix)
+    else:
+        from cuml.metrics import pairwise_distances
+
+        precomputed_knn = pairwise_distances(X, metric=knn_metric)
+        assert isinstance(precomputed_knn, cp.ndarray)
+
+    with CleanSparkSession() as spark:
+        pyspark_type = "float"
+        feature_cols = [f"c{i}" for i in range(X.shape[1])]
+        schema = [f"{c} {pyspark_type}" for c in feature_cols]
+        df = spark.createDataFrame(X.tolist(), ",".join(schema))
+        df = df.withColumn("features", array(*feature_cols)).drop(*feature_cols)
+
+        umap = UMAP(
+            num_workers=gpu_number,
+            metric="sqeuclidean",
+            random_state=random_state,
+            precomputed_knn=precomputed_knn,
```

Comment:
Can we convert back to numpy to emphasize that numpy ok here in keeping with our no need for gpu on driver?